### PR TITLE
[Community Pipeline] UnCLIP Text Interpolation Pipeline

### PR DESCRIPTION
**Pull request by Abhinay1997**

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
This looks super cool, thanks for adding it @Abhinay1997 

Would be interesting to try it out on some of the examples of DALLE-2, e.g.

"a photo of an adult lion → a photo of lion cub"
"a photo of a landscape in winter → a photo of a landscape in fall"
"a photo of a victorian house → a photo of a modern house"

Also think this is a cool pipeline to build a spaces with :-) 

@williamberman can you also have a look here?

**Comment by Abhinay1997** [comment]
Thanks @patrickvonplaten. 😄

Need your input on the attention mask to be used for the interpolated text embeddings because the results are not great when the difference in prompt length is large.

**Comment by patrickvonplaten** [comment]
> Thanks @patrickvonplaten. smile
> 
> Need your input on the attention mask to be used for the interpolated text embeddings because the results are not great when the difference in prompt length is large.

cc @williamberman maybe?

**Comment by Abhinay1997** [comment]
@patrickvonplaten , see our discussion here #1869 willberman suggested we use the larger of the two for now

File: examples/community/unclip_text_interpolation.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+                generator=generator,
+                prior_num_inference_steps=prior_num_inference_steps,
+                decoder_num_inference_steps=decoder_num_inference_steps,
+                super_res_num_inference_steps=super_res_num_inference_steps,
+                prior_guidance_scale=prior_guidance_scale,
+                decoder_guidance_scale=decoder_guidance_scale,
+                output_type=output_type,
+                return_dict=return_dict,
+            )
+
```
Comment (williamberman) [inline examples/community/unclip_text_interpolation.py:667]: I think ideally we should batch the embeddings instead of effectively running the pipeline in a loop

File: examples/community/unclip_text_interpolation.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+                text_embeds = end_text_embeds
+                last_hidden_state = end_last_hidden_state
+            else:
+                text_embeds = UnCLIPTextInterpolationPipeline.slerp(interp_val, start_text_embeds, end_text_embeds)
+                last_hidden_state = UnCLIPTextInterpolationPipeline.slerp(
+                    interp_val, start_last_hidden_state, end_last_hidden_state
+                )
+
+            text_model_output.text_embeds = text_embeds.unsqueeze(0).to(device)
+            text_model_output.last_hidden_state = last_hidden_state.unsqueeze(0).to(device)
```
Comment (williamberman) [inline examples/community/unclip_text_interpolation.py:653]: ideally we use the interpolated results directly instead of mutating `text_model_output`

File: examples/community/unclip_text_interpolation.py [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+        if output_type == "pil":
+            image = self.numpy_to_pil(image)
+
+        if not return_dict:
+            return (image,)
+
+        return ImagePipelineOutput(images=image)
+
+    @staticmethod
+    def slerp(val, low, high):
```
Comment (williamberman) [inline examples/community/unclip_text_interpolation.py:535]: nice! slerp doesn't have to be a static or regular method on the class. Let's just move it to a regular function at the top of the file :)

File: examples/community/unclip_text_interpolation.py [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+                hasattr(module, "_hf_hook")
+                and hasattr(module._hf_hook, "execution_device")
+                and module._hf_hook.execution_device is not None
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
+
+    @torch.no_grad()
+    # Copied from diffusers.pipelines.unclip.pipeline_unclip.UnCLIPPipeline.__call__
+    def _generate(
```
Comment (williamberman) [inline examples/community/unclip_text_interpolation.py:250]: We try to keep the `__call__` function pretty self contained so lets move `_generate` back to inside `__call__`. This should work well with the other comment on batching the interpolated text embeddings :)

**Comment by williamberman** [comment]
Great start @Abhinay1997 !

File: examples/community/unclip_text_interpolation.py (hunk 2 above)
Comment (Abhinay1997) [reply examples/community/unclip_text_interpolation.py:653]: Got it. Will make the change.

File: examples/community/unclip_text_interpolation.py (hunk 3 above)
Comment (Abhinay1997) [reply examples/community/unclip_text_interpolation.py:535]: Ohh. Yeah that makes sense.

File: examples/community/unclip_text_interpolation.py (hunk 4 above)
Comment (Abhinay1997) [reply examples/community/unclip_text_interpolation.py:250]: Do you mean like this ?

```python
def __call__(.....):
     def _generate(.....):
```

File: examples/community/unclip_text_interpolation.py (hunk 1 above)
Comment (Abhinay1997) [reply examples/community/unclip_text_interpolation.py:667]: Sure. Will batch the pipeline run.

**Comment by Abhinay1997** [comment]
@williamberman , made changes based on your feedback. Could you review them when you can ?

File: examples/community/unclip_text_interpolation.py [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+        for interp_val in torch.linspace(0, 1, steps):
+            text_embeds = slerp(interp_val, text_model_output.text_embeds[0], text_model_output.text_embeds[1])
+            last_hidden_state = slerp(
+                interp_val, text_model_output.last_hidden_state[0], text_model_output.last_hidden_state[1]
+            )
+            batch_text_embeds.append(text_embeds.unsqueeze(0))
+            batch_last_hidden_state.append(last_hidden_state.unsqueeze(0))
+
+        batch_text_embeds = torch.cat(batch_text_embeds)
+        batch_last_hidden_state = torch.cat(batch_last_hidden_state)
```
Comment (williamberman) [inline examples/community/unclip_text_interpolation.py:650]: nice!

File: examples/community/unclip_text_interpolation.py (hunk 4 above)
Comment (williamberman) [reply examples/community/unclip_text_interpolation.py]: almost! could we just remove the `_generate` function and have all of the logic directly in the `__call__` method?

File: examples/community/unclip_text_interpolation.py (hunk 2 above)
Comment (williamberman) [reply examples/community/unclip_text_interpolation.py]: > @williamberman , made changes based on your feedback. Could you review them when you can ?
> 
> P.S Ran the code through black and isort multiple times but it's still failing the code quality test

We recently updated the versions of our linters etc.. could you try making sure they're up to date and running `make style` locally before pushing?

**Comment by williamberman** [comment]
Love the progress @Abhinay1997 ! Could you also add some example code for running the pipeline along with the outputs it gives :) ?

File: examples/community/unclip_text_interpolation.py (hunk 4 above)
Comment (Abhinay1997) [reply examples/community/unclip_text_interpolation.py]: Sure !

File: examples/community/README.md [hunk 6]
```diff
@@ -951,3 +953,39 @@ print(pipeline.prior_scheduler)
...
+```
+
+The resulting images in order:-
+
+![result_0](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_0.png)
+![result_1](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_1.png)
+![result_2](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_2.png)
+![result_3](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_3.png)
+![result_4](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_4.png)
+![result_5](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_5.png)
```
Comment (Abhinay1997) [inline examples/community/README.md:991]: @williamberman Code example for the pipeline.

**Comment by Abhinay1997** [comment]
@williamberman 

1. Refactored the __call__ function as suggested :) 
2. I updated to the same linter versions as the github action and `make style` works on my local machine but fails the check for the PR.

File: examples/community/unclip_text_interpolation.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+    UNet2DConditionModel,
+    UNet2DModel,
+)
+from diffusers.pipelines.unclip import UnCLIPTextProjModel
+from diffusers.utils import is_accelerate_available, logging, randn_tensor
+from transformers import CLIPTextModelWithProjection, CLIPTokenizer
+from transformers.models.clip.modeling_clip import CLIPTextModelOutput
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
```
Comment (williamberman) [inline examples/community/unclip_text_interpolation.py:21]:
```suggestion
import torch
from torch.nn import functional as F
from transformers import CLIPTextModelWithProjection, CLIPTokenizer
from transformers.models.clip.modeling_clip import CLIPTextModelOutput

from diffusers import (
    DiffusionPipeline,
    ImagePipelineOutput,
    PriorTransformer,
    UnCLIPScheduler,
    UNet2DConditionModel,
    UNet2DModel,
)
from diffusers.pipelines.unclip import UnCLIPTextProjModel
from diffusers.utils import is_accelerate_available, logging, randn_tensor


logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
```

Comment (williamberman) [reply examples/community/unclip_text_interpolation.py:21]: Should help it pass CI :)

**Comment by williamberman** [comment]
Awesome, looks basically good to do @Abhinay1997! I needed to merge in master to get the updated linter versions :)

**Comment by Abhinay1997** [comment]
Thanks for the help Will ! Hope we are good for the merge now.

**Review by williamberman — APPROVED** [review]

**Comment by williamberman** [comment]
Awesome, this is great @Abhinay1997! Would you be interested in making a spaces to showcase the pipeline? https://huggingface.co/spaces

**Comment by Abhinay1997** [comment]
Sure @williamberman ! I was thinking of doing it once the PR is merged :)
